lib/xain/helpers.ex

```diff
@@ -37,6 +38,18 @@ defmodule Xain.Helpers do
   def pop!([item | rest]) do
     {rest, item |> Enum.reverse}
   end
+
+  def ensure_valid_contents(contents, _) when is_binary(contents) or is_list(contents) do
+    contents
+  end
+  def ensure_valid_contents(contents, tag_name) when is_number(contents) or is_atom(contents) do
+    Logger.debug "#{tag_name} has been called as #{tag_name}(#{inspect(contents)}, ...), but the first argument supposed to be a binary"
+    to_string(contents)
+  end
+  def ensure_valid_contents(contents, tag_name) do
+    Logger.warn "#{tag_name} has been called as #{tag_name}(#{inspect(contents)}, ...), but the first argument supposed to be a binary"
+    inspect(contents)
```

Comment:
@smpallen99 Wow, you have already started ex_admin update.. Cool! I was planning to work on it on the weekend )))

> One of the issues I’m facing is with the safe callback being run for each do markup.

I like your idea with `safe: true`. Also, please, note that you could use returning of a list or join with `<>` instead of `markup` in some cases .
For example:
``` elixir
  def build_item(conn, %{type: :input, resource: _resource, name: field_name, opts: opts},
       resource, model_name, errors) do
    # IO.puts "build_item 4: #{inspect field_name}"
    errors = get_errors(errors, field_name)
    label = get_label(field_name, opts)
    required = if field_name in (conn.assigns[:ea_required] || []), do: true, else: false
    wrap_item(resource, field_name, model_name, label, errors, opts, conn.params, required, fn(ext_name) ->
      field_type = resource.__struct__.__schema__(:type, field_name)
      [
        build_control(field_type, resource, opts, model_name, field_name, ext_name),
        build_errors(errors)
      ]
    end)
  end
```
instead of markup block in each `build_control` clause.